haystack_experimental/components/tools/openapi/_openapi.py

```diff
@@ -0,0 +1,366 @@
+        openapi_spec: Union[str, Path],
+        credentials: Optional[str] = None,
+        request_sender: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
+        llm_provider: Optional[LLMProvider] = None,
+    ):  # noqa: PLR0913
+        """
+        Initialize a ClientConfiguration instance.
+
+        :param openapi_spec: The OpenAPI specification as a file path, URL, or dictionary.
+        :param credentials: The credentials to use for authentication.
+        :param request_sender: The function to use for sending requests.
+        :param llm_provider: The LLM provider to use for generating tools definitions.
+        :raises ValueError: If the OpenAPI specification format is invalid.
+        """
+        if isinstance(openapi_spec, (str, Path)) and os.path.isfile(openapi_spec):
+            self.openapi_spec = OpenAPISpecification.from_file(openapi_spec)
+        elif isinstance(openapi_spec, str):
+            if is_valid_http_url(openapi_spec):
+                self.openapi_spec = OpenAPISpecification.from_url(openapi_spec)
+            else:
+                self.openapi_spec = OpenAPISpecification._from_str(openapi_spec)
+        else:
+            raise ValueError(
+                "Invalid OpenAPI specification format. Expected file path or dictionary."
+            )
+
+        self.credentials = credentials
+        self.request_sender = request_sender or send_request
+        self.llm_provider: LLMProvider = llm_provider or LLMProvider.OPENAI
+
+    def get_auth_function(self) -> Callable[[Dict[str, Any], Dict[str, Any]], Any]:
+        """
+        Get the authentication function that sets a schema specified authentication to the request.
+
+        The function takes a security scheme and a request as arguments:
+            `security_scheme: Dict[str, Any] - The security scheme from the OpenAPI spec.`
+            `request: Dict[str, Any] - The request to apply the authentication to.`
+        :returns: The authentication function.
+        :raises ValueError: If the credentials type is not supported.
+        """
+        security_schemes = self.openapi_spec.get_security_schemes()
+        if not self.credentials:
+            return lambda security_scheme, request: None  # No-op function
+        if isinstance(self.credentials, str):
+            return self._create_authentication_from_string(
+                self.credentials, security_schemes
+            )
+        raise ValueError(f"Unsupported credentials type: {type(self.credentials)}")
+
+    def get_tools_definitions(self) -> List[Dict[str, Any]]:
+        """
+        Get the tools definitions used as tools LLM parameter.
+
+        :returns: The tools definitions passed to the LLM as tools parameter.
+        """
+        provider_to_converter = defaultdict(
+            lambda: openai_converter,
+            {
+                LLMProvider.ANTHROPIC.value: anthropic_converter,
+                LLMProvider.COHERE.value: cohere_converter,
```

Comment:
We can just keep this as `Dict[LLMProvider, Callable]`:
```suggestion
                LLMProvider.ANTHROPIC: anthropic_converter,
                LLMProvider.COHERE: cohere_converter,
```